### PR TITLE
8.1.2.1. Pseudo-Header Fields: Sends a HEADERS frame that contains the pseudo-header field defined for response

### PR DESCRIPTION
**Pull request by joedevivo**
```
      8.1.2.1. Pseudo-Header Fields
        × Sends a HEADERS frame that contains the pseudo-header field defined for response
          - The endpoint MUST respond with a stream error of type PROTOCOL_ERROR.
            Expected: GOAWAY frame (ErrorCode: PROTOCOL_ERROR)
                      RST_STREAM frame (ErrorCode: PROTOCOL_ERROR)
                      Connection close
              Actual: DATA frame (Length: 16, Flags: 1)
```

### CODE REVIEW

**Comment by joedevivo** [comment]
This PR only allows requests to contain the pseudo headers `:path`, `:authority`, `:method`, and `:scheme`

File: src/http2_connection.erl [hunk 1]
```diff
@@ -1476,13 +1476,24 @@ maybe_hpack(Continuation, Conn)
...
 
+    HeadersType =
+        case {Continuation#continuation_state.type,
+              Conn#connection.type
+             } of
+            {headers, server} ->
+                request;
+            {push_promise, client} ->
+                request;
+             _ -> response
```
Comment (stillinbeta) [inline src/http2_connection.erl]: do you need to do seperate validation for trailers?

File: src/http2_connection.erl [hunk 2]
```diff
@@ -1529,10 +1540,9 @@ maybe_hpack(Continuation, Conn) ->
 %% * Only acceptable psuedoheaders are:
 %%     :method, :scheme, :authority, :path,
 
-good_req_headers(Headers) ->
+good_headers(Type, Headers) ->
```
Comment (stillinbeta) [inline src/http2_connection.erl]: not sure I get what this function's for?

Comment (joedevivo) [inline src/http2_connection.erl]: It's *THE* function. "Is this set of headers valid or not?"

File: src/http2_connection.erl (hunk 1 above)
Comment (joedevivo) [inline src/http2_connection.erl]: Trailers will get further attention in #43

File: src/http2_connection.erl [hunk 3]
```diff
@@ -1529,10 +1540,9 @@ maybe_hpack(Continuation, Conn) ->
 %% * Only acceptable psuedoheaders are:
 %%     :method, :scheme, :authority, :path,
```
Comment (stillinbeta) [inline src/http2_connection.erl]: update this comment maybe?

Comment (joedevivo) [inline src/http2_connection.erl]: Already in progress :D

File: src/http2_connection.erl (hunk 2 above)
Comment (stillinbeta) [inline src/http2_connection.erl]: so are you planning to add things that aren't validate_pseudos later?

Comment (joedevivo) [inline src/http2_connection.erl]: Originally it was something like

```erlang
validate_pseudos(Headers) andalso no_upper_names(Headers)
```

but I squeaked a little performance optimization out by only running no_upper_names on Headers' tail after the pseudos were already looked at.

This function effectively frees me to write true/false validators for the Headers list and wrap it up in a compression error for the `hpack_maybe` function.

File: src/http2_connection.erl [hunk 4]
```diff
@@ -1541,22 +1562,25 @@ good_req_headers(Headers) ->
...
-validate_pseudos([{<<":path">>,_V}|Tail]) ->
-    validate_pseudos(Tail);
-validate_pseudos([{<<":method">>,_V}|Tail]) ->
-    validate_pseudos(Tail);
-validate_pseudos([{<<":scheme">>,_V}|Tail]) ->
-    validate_pseudos(Tail);
-validate_pseudos([{<<":authority">>,_V}|Tail]) ->
-    validate_pseudos(Tail);
-validate_pseudos(DoneWithPseudos) ->
+validate_pseudos(request, [{<<":path">>,_V}|Tail]) ->
```
Comment (stillinbeta) [inline src/http2_connection.erl]: it looks like you don't validate the presence of the required headers, just reject invalid pseudos / wrong order?

Comment (joedevivo) [inline src/http2_connection.erl]: correct. This test, and #40 which this also fixes, are only about failing when incorrect pseudoheaders are encountered.

Actually, it doesn't seem like the `h2spec` tool tests for "missing required pseudoheader", so I'll make note to figure out what that means and include another test for it.

**Comment by stillinbeta** [comment]
+1 shipit

**Pull request merged**